### ISSUE TITLE
[BMG] PyTorch 2.12 performance cases where inductor is slower than eager (speedup < 1)

### ISSUE THREAD

**Issue by mengfei25** [intel/torch-xpu-ops#3761]
## Summary

BMG PyTorch 2.12 release performance results contain multiple cases where inductor is slower than eager (`speedup < 1`).

These are PyTorch 2.12 release performance results collected from the raw performance artifacts:

- Platform: BMG (Battlemage)
- PyTorch: `torch 2.12.0+xpu`
- Environment: Ubuntu 24.04 LTS, `intel-opencl-icd 26.05.37020.3`, `libze-dev 1.27.0`, `intel-opencl-rt 2025.3.2`, `oneMKL 2025.3.1`

## Representative Cases

The rows below come directly from the raw PyTorch 2.12 performance CSVs and show `speedup < 1`, which means inductor is slower than eager on the same workload.

| suite | dtype | mode | model | batch_size | speedup | source csv |
| --- | --- | --- | --- | ---: | ---: | --- |
| huggingface | bfloat16 | inference | `ElectraForQuestionAnswering` | 128 | 0.662224 | `huggingface/bfloat16/inference/performance/inductor-results-huggingface-bfloat16-inference-xpu-performance.csv` |
| huggingface | bfloat16 | inference | `BertForQuestionAnswering` | 32 | 0.732601 | `huggingface/bfloat16/inference/performance/inductor-results-huggingface-bfloat16-inference-xpu-performance.csv` |
| huggingface | bfloat16 | inference | `BartForConditionalGeneration` | 4 | 0.741895 | `huggingface/bfloat16/inference/performance/inductor-results-huggingface-bfloat16-inference-xpu-performance.csv` |
| huggingface | bfloat16 | inference | `MBartForConditionalGeneration` | 4 | 0.723328 | `huggingface/bfloat16/inference/performance/inductor-results-huggingface-bfloat16-inference-xpu-performance.csv` |
| huggingface | bfloat16 | inference | `PLBartForConditionalGeneration` | 8 | 0.829213 | `huggingface/bfloat16/inference/performance/inductor-results-huggingface-bfloat16-inference-xpu-performance.csv` |
| huggingface | amp_bf16 | inference | `BertForQuestionAnswering` | 32 | 0.893284 | `huggingface/amp_bf16/inference/performance/inductor-results-huggingface-amp_bf16-inference-xpu-performance.csv` |
| huggingface | float16 | inference | `ElectraForQuestionAnswering` | 128 | 0.646220 | `huggingface/float16/inference/performance/inductor-results-huggingface-float16-inference-xpu-performance.csv` |
| huggingface | bfloat16 | training | `OPTForCausalLM` | 2 | 0.876858 | `huggingface/bfloat16/training/performance/inductor-results-huggingface-bfloat16-training-xpu-performance.csv` |
| torchbench | bfloat16 | inference | `lennard_jones` | 1000 | 0.671808 | `torchbench/bfloat16/inference/performance/inductor-results-torchbench-bfloat16-inference-xpu-performance.csv` |
| torchbench | bfloat16 | inference | `maml` | 1 | 0.664822 | `torchbench/bfloat16/inference/performance/inductor-results-torchbench-bfloat16-inference-xpu-performance.csv` |
| torchbench | bfloat16 | inference | `soft_actor_critic` | 256 | 0.865522 | `torchbench/bfloat16/inference/performance/inductor-results-torchbench-bfloat16-inference-xpu-performance.csv` |

There are additional PyTorch 2.12 rows with the same pattern outside this representative subset.

## Notes

- The pattern is concentrated in inference workloads, especially HuggingFace QA/seq2seq style models and several TorchBench models.
- These rows were taken from the raw perf artifacts rather than the generated comparison CSV, because the goal here is to flag inductor-vs-eager gaps directly in the PyTorch 2.12 release results.

## Ask

Please review these BMG PyTorch 2.12 cases and track the inductor performance gaps where eager is still faster than inductor.

**Comment by msnliu** [comment]
speedup reproduction:
<img width="614" height="240" alt="Image" src="https://github.com/user-attachments/assets/2d220242-5a5c-4db5-b8a5-dcd22324d23a" />

device ratio:
<img width="902" height="256" alt="Image" src="https://github.com/user-attachments/assets/6b18df74-edd5-48a5-9da6-b2f2aa3e837c" />

Env:
| Component | `torch-2.12-xpu` | `torch-2.14-xpu` | Notes |
| :--- | :--- | :--- | :--- |
| **Python** | `3.13.14` | `3.13.15` | Aligned |
| **PyTorch** | `2.12.0+xpu` | `2.14.0+xpu` | Official release wheels |
| **Triton** | `3.7.1` | `3.8.0` | Default bundled versions |
| **Transformers** | `5.14.1` | `5.14.1` | Aligned |
| **TorchBench** | `Installed` | `Installed` | Aligned |

**Comment by msnliu** [comment]
mainly investigated 08_opt_bf16_train (the only case whose speedup dropped from torch 2.12 to 2.14)
benchmark:

<img width="759" height="348" alt="Image" src="https://github.com/user-attachments/assets/1e2496df-6f11-429b-bc12-16d3083d6def" />
 
Compared the performances on XPU with Cuda:
 
<img width="663" height="372" alt="Image" src="https://github.com/user-attachments/assets/ab4f1d19-1657-4113-9e78-8193c60d848f" />

Investigated the 3 regressed/related XPU cases (08_opt_bf16_train, 01_electra_bf16_inf, 07_electra_fp16_inf). torch.profile shows all 3 are kernel-bound: the speedup is capped by GPU compute.

Graph breaks:
Verified the OPT LayerDrop sync point via a `modeling_opt.py` patch: removing the data-dependent graph break (`item()` sync) improved Inductor speedup by +4.6% on XPU and +3.5% on CUDA. This accounts for only ~4%, so it is not the main bottleneck — the remaining <1.0 speedup is driven by high kernel time.

**Comment by msnliu** [comment]
In PyTorch 2.12, models with 4D broadcasted attention masks (e.g. HuggingFace BERT/BART/Electra with shape [B, 1, 1, S]) fail to dispatch to the oneDNN SDPA backend (overrideable_attention) on XPU, falling back to the un-fused math backend.

Profiling traces show that this fallback introduces extra general GEMMs and unfused softmax kernels per attention block, causing substantial performance regressions (speedup < 1.0). 

after the sdpa path fix (https://github.com/pytorch/pytorch/pull/197505):
<img width="624" height="241" alt="Image" src="https://github.com/user-attachments/assets/d67f7597-e2bd-46a7-991d-b38dde040460" />
we could see that model 05_plbart_bf16_inf and 06_bert_amp_inf did not show regression again:
- 05_plbart_bf16_inf:  121 gemm + 12 triton safe_softmax_add + 6 micro_sdpa -> 97 gemm + 18 micro_sdpa
- 06_bert_amp_inf: 97 gemm + 12 triton safe_softmax_add -> 73 gemm + 12 micro_sdpa
- and all other models that have attention blocks improved (models that don't have attention: 09_lennard_bf16_inf, 10_maml_bf16_inf, 11_sac_bf16_inf)

**Comment by msnliu** [comment]
Transformers issues: 
https://github.com/huggingface/transformers/issues/48924
https://github.com/huggingface/transformers/issues/48925
After applying the Transformers masking fix (https://github.com/huggingface/transformers/pull/48975), all 8 validation cases pass successfully: Electra, BERT, BART, mBART, PLBART, BERT AMP, Electra FP16, and OPT BF16 training. The fix correctly enables mask skipping under torch.compile. 

<img width="672" height="184" alt="Image" src="https://github.com/user-attachments/assets/38784702-0e64-4481-960e-d2497a396679" />

**Comment by msnliu** [comment]
MAML inference performs task-specific finetuning and calls `torch.autograd.grad()`. With PyTorch 2.14's default Dynamo settings, autograd operations are not traced, which causes graph breaks.

This change:
- enables `trace_autograd_ops` for MAML;
- passes parameters and BatchNorm state explicitly;
- removes `deepcopy()` and `Tensor.item()` from the compiled path.

MAML inference now compiles without the observed graph breaks or recompilations. Validated on Intel XPU with numerical correctness checks and approximately 2.7x Inductor speedup over eager.

https://github.com/pytorch/benchmark/pull/2709
